keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,762 @@
+
+[summary-pod-stats-object]: https://github.com/kubernetes/kubernetes/blob/release-1.20/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go#L102-L132
+
+### User Stories [optional]
+
+#### Story 1
+As a cluster admin, I would like my node to be as performant as possible, and not waste resources by duplicating stats collection by CRI and cAdvisor.
+
+#### Story 2
+As a node maintainer, I would like to clarify the sources of truth for each stat the Kubelet reports to the Summary API, so that I can better find which component is to investigate for any issues.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### History/Past Conversations
+
+There have been conversations in the past on [deprecating the stats Summary API](#core-proposal), in favor of allowing an on-demand model of stats gathering/reporting.
+This KEP does not attempt to implement that. These stats are largely considered [a part of Kubernetes](#keep-summary). The stated goal of this KEP is to reduce the overhead
+on two entities reporting metrics, not totally changing what stats the Kubelet reports (with the exception of Accelerator and UserDefinedMetrics).
+
+Thus, this KEP largely the plan described [here](#plan), with some changes:
+
+- the `/metrics/cadvisor` endpoint will likely not be exposed in the Alpha release, and its continued support will need to be evaluated for the future.
+- CRI is used for all of the monitoring endpoints related to Containers and Pods (except Volume and Ephemeral Storage)
+- CRI is used to provide metrics for eviction.
+
+[core-proposal]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/core-metrics-pipeline.md#proposed-core-metrics
+[keep-summary]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-666636130
+[plan]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-724928827
+
+#### cAdvisor Metrics Endpoint
+
+In addition to the `/stats/summary` endpoint, there are also consumers that rely on `/metrics/cadvisor`.
+In the past, each of these were populated primarily by cAdvisor, but the goals of this KEP will prevent cAdvisor from hosting these metrics.
+Where the majority of internal Kubernetes entities rely solely on the Summary API, there are likely consumers in the community that rely on this endppoint.
+
+Thus, in addition to fulfilling `/stats/summary` endpoint, this KEP will have to decide on how to handle `/metrics/cadvisor`
+Below is the proposed strategy for doing so:
+
+1. For the Alpha release of this KEP, accept there will be breaking changes to `/metrics/cadvisor`. Use a feature gate and the strategies proposed above to minimize breakages.
+2. For the Beta release, evaluate the future of the `/metrics/cadvisor` endpoint. We can either:
+    - Deprecate the endpoint, and extend the Summary API to report the required fields.
+    - Move the reporter of `/metrics/cadvisor` from cAdvisor to the CRI implementation.
```

Comment:
Does the `/metrics/cadvisor` endpoint need to be implemented in the CRI implementation, or we just need equivalent metrics there?

Reply:
Someone needs to support /metrics/cadvisor, and the basis of this KEP is that it isn't cadvisor. The best place is the CRI impl directly, IMO

Reply:
But it is not clear to me whether we are asking CRI runtimes to support  the `/metrics/cadvisor` endpoint, or just provide equivalent information in another endpoint.

It feels weird to ask the container runtime to support a `cadvisor` endpoint... And containerd already has a metrics endpoint exposing all container cgroup stats today.